lightning/src/routing/scoring.rs

```diff
@@ -460,25 +466,420 @@ impl<T: Time> Readable for ChannelFailure<T> {
 	}
 }
 
+/// [`Score`] implementation using channel success probability distributions.
+///
+/// Based on *Optimally Reliable & Cheap Payment Flows on the Lightning Network* by Rene Pickhardt
+/// and Stefan Richter [[1]]. Given the uncertainty of channel liquidity balances, probability
+/// distributions are defined based on knowledge learned from successful and unsuccessful attempts.
+/// Then the negative log of the success probability is used to determine the cost of routing a
+/// specific HTLC amount through a channel.
+///
+/// Knowledge about channel liquidity balances takes the form of upper and lower bounds on the
+/// possible liquidity. Certainty of the bounds is decreased over time using a decay function. See
+/// [`ProbabilisticScoringParameters`] for details.
+///
+/// [1]: https://arxiv.org/abs/2107.05322
+pub type ProbabilisticScorer<G> = ProbabilisticScorerUsingTime::<G, ConfiguredTime>;
+
+/// Probabilistic [`Score`] implementation.
+///
+/// # Note
+///
+/// Mixing the `no-std` feature between serialization and deserialization results in undefined
+/// behavior.
+///
+/// (C-not exported) generally all users should use the [`ProbabilisticScorer`] type alias.
+#[doc(hidden)]
+pub struct ProbabilisticScorerUsingTime<G: Deref<Target = NetworkGraph>, T: Time> {
+	params: ProbabilisticScoringParameters,
+	network_graph: G,
+	// TODO: Remove entries of closed channels.
+	channel_liquidities: HashMap<u64, ChannelLiquidity<T>>,
+}
+
+/// Parameters for configuring [`ProbabilisticScorer`].
+#[derive(Clone, Copy)]
+pub struct ProbabilisticScoringParameters {
+	/// A multiplier used to determine the amount in msats willing to be paid to avoid routing
+	/// through a channel, as per multiplying by the negative log of the channel's success
+	/// probability for a payment.
+	///
+	/// The success probability is determined by the effective channel capacity, the payment amount,
+	/// and knowledge learned from prior successful and unsuccessful payments. The lower bound of
+	/// the success probability is 0.01, effectively limiting the penalty to the range
+	/// `0..=2*liquidity_penalty_multiplier_msat`. The knowledge learned is decayed over time based
+	/// on [`liquidity_offset_half_life`].
+	///
+	/// Default value: 1,000 msat
```

Comment:
Hmm, 1000 feels very low, the defaults for the current scorer let us get up to ~18sats in total penalty, whereas this will max out at 2. 2 intuitively feels like it could often not be enough to heavily de-prioritize a channel that keeps failing, given lots of nodes charge a 1sat base fee. I'd guess something closer to 10 sats is probably where this should be, that would give us a max of 20 sats penalty, with 10 sats penalty for a 10% chance of success, 3 sats for 50% chance, and 0.45 sats for a 90% chance of success.

Reply:
Yeah, that seems like a better default.